consensus/src/memory_pool.rs

```diff
@@ -18,419 +18,270 @@
+            transaction,
+        });
 
         Ok(Some(transaction_id))
     }
 
     /// Cleanse the memory pool of outdated transactions.
-    #[inline]
-    pub async fn cleanse<P: LoadableMerkleParameters, S: Storage>(
-        &self,
-        storage: &Ledger<T, P, S>,
-    ) -> Result<(), ConsensusError> {
-        let new_memory_pool = Self::new();
-
-        for (_, entry) in self.clone().transactions.inner().iter() {
-            new_memory_pool.insert(storage, entry.clone()).await?;
-        }
+    pub fn cleanse(&self, ledger: &DynLedger) -> Result<MemoryPool, ConsensusError> {
+        let mut new_pool = Self::new();
 
-        self.total_size_in_bytes.store(
-            new_memory_pool.total_size_in_bytes.load(Ordering::SeqCst),
-            Ordering::SeqCst,
-        );
-        self.transactions.reset(new_memory_pool.transactions.inner_full()).await;
+        for (_, entry) in &self.transactions {
+            new_pool.insert(ledger, entry.transaction.clone())?;
+        }
 
-        Ok(())
+        Ok(new_pool)
     }
 
-    /// Removes transaction from memory pool or error.
-    #[inline]
-    pub async fn remove(&self, entry: &Entry<T>) -> Result<Option<Vec<u8>>, ConsensusError> {
-        if self.contains(entry) {
-            self.total_size_in_bytes
-                .fetch_sub(entry.size_in_bytes, Ordering::SeqCst);
+    // /// Removes transaction from memory pool or error.
+    // #[inline]
+    // pub async fn remove(&self, entry: &Entry) -> Result<Option<Vec<u8>>, ConsensusError> {
+    //     if self.contains(entry) {
+    //         let transaction_id = entry.transaction.transaction_id()?.to_vec();
 
-            let transaction_id = entry.transaction.transaction_id()?.to_vec();
+    //         self.transactions.remove(transaction_id.to_vec()).await;
 
-            self.transactions.remove(transaction_id.to_vec()).await;
-
-            return Ok(Some(transaction_id));
-        }
+    //         return Ok(Some(transaction_id));
+    //     }
 
-        Ok(None)
-    }
+    //     Ok(None)
+    // }
 
     /// Removes transaction from memory pool based on the transaction id.
-    #[inline]
-    pub async fn remove_by_hash(&self, transaction_id: &[u8]) -> Result<Option<Entry<T>>, ConsensusError> {
-        match self.transactions.get(transaction_id) {
+    pub fn remove(&mut self, transaction_id: &Digest) -> Result<Option<SerialTransaction>, ConsensusError> {
+        match self.transactions.remove(transaction_id) {
             Some(entry) => {
-                self.total_size_in_bytes
-                    .fetch_sub(entry.size_in_bytes, Ordering::SeqCst);
-
-                self.transactions.remove(transaction_id.to_vec()).await;
-
-                Ok(Some(entry.clone()))
+                for commitment in &entry.transaction.new_commitments {
+                    if !self.commitments.remove(commitment) {
+                        panic!("missing commitment from memory pool during removal");
```

Comment:
we're in a method returning a `Result`; is the severity of the issue (presumably a logic bug) sufficient to `panic`?